File: backend/app/schemas/kodee.py
"""
Kodee NL-to-SQL Schemas

Defines data models for the Kodee natural language to SQL engine,
including schema context, query generation, and chat interfaces.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class QueryIntent(str, Enum):
    """Detected intent from natural language query"""
    SELECT = "select"
    AGGREGATE = "aggregate"
    COMPARE = "compare"
    TREND = "trend"
    TOP_N = "top_n"
    FILTER = "filter"
    JOIN = "join"
    UNKNOWN = "unknown"


class QueryComplexity(str, Enum):
    """Complexity level of the generated query"""
    SIMPLE = "simple"
    MODERATE = "moderate"
    COMPLEX = "complex"


class MessageRole(str, Enum):
    """Role in chat conversation"""
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class ValidationStatus(str, Enum):
    """SQL validation status"""
    VALID = "valid"
    WARNING = "warning"
    INVALID = "invalid"
    BLOCKED = "blocked"


# Schema Context Models

class ColumnInfo(BaseModel):
    """Information about a database column"""
    name: str
    data_type: str
    nullable: bool = True
    description: Optional[str] = None
    sample_values: list[Any] = Field(default_factory=list)
    is_primary_key: bool = False
    is_foreign_key: bool = False
    foreign_key_ref: Optional[str] = None  # "table.column"


class TableInfo(BaseModel):
    """Information about a database table"""
    name: str
    schema_name: str = "public"
    description: Optional[str] = None
    columns: list[ColumnInfo] = Field(default_factory=list)
    row_count: Optional[int] = None
    aliases: list[str] = Field(default_factory=list)


class RelationshipInfo(BaseModel):
    """Information about table relationships"""
    from_table: str
    from_column: str
    to_table: str
    to_column: str
    relationship_type: str  # "one_to_one", "one_to_many", "many_to_many"


class MeasureInfo(BaseModel):
    """Information about semantic model measures"""
    id: str
    name: str
    description: Optional[str] = None
    expression: str  # e.g., "SUM(sales.amount)"
    table: str
    column: str
    aggregate: str


class DimensionInfo(BaseModel):
    """Information about semantic model dimensions"""
    id: str
    name: str
    description: Optional[str] = None
    table: str
    column: str
    hierarchy: Optional[list[str]] = None


class SchemaContext(BaseModel):
    """Complete schema context for NL-to-SQL generation"""
    connection_id: str
    model_id: Optional[str] = None
    model_name: Optional[str] = None
    tables: list[TableInfo] = Field(default_factory=list)
    relationships: list[RelationshipInfo] = Field(default_factory=list)
    measures: list[MeasureInfo] = Field(default_factory=list)
    dimensions: list[DimensionInfo] = Field(default_factory=list)
    dialect: str = "postgresql"  # postgresql, mysql, bigquery, snowflake
    additional_context: Optional[str] = None  # User-provided context


# Query Request/Response Models

class NLQueryRequest(BaseModel):
    """Request for NL-to-SQL conversion"""
    question: str = Field(..., description="Natural language question")
    connection_id: Optional[str] = None
    model_id: Optional[str] = None
    conversation_id: Optional[str] = None
    include_explanation: bool = True
    max_rows: int = Field(default=1000, le=10000)
    timeout_seconds: int = Field(default=30, le=120)


class QueryValidation(BaseModel):
    """SQL validation result"""
    status: ValidationStatus
    messages: list[str] = Field(default_factory=list)
    blocked_keywords: list[str] = Field(default_factory=list)
    estimated_complexity: Optional[QueryComplexity] = None
    estimated_rows: Optional[int] = None


class NLQueryResponse(BaseModel):
    """Response from NL-to-SQL conversion"""
    sql: str
    explanation: str
    confidence: float = Field(..., ge=0, le=1)
    intent: QueryIntent = QueryIntent.UNKNOWN
    complexity: QueryComplexity = QueryComplexity.SIMPLE
    validation: QueryValidation
    tables_used: list[str] = Field(default_factory=list)
    columns_used: list[str] = Field(default_factory=list)
    alternatives: list[str] = Field(default_factory=list)  # Alternative SQL queries
    follow_up_questions: list[str] = Field(default_factory=list)
    execution_time_ms: Optional[float] = None
    conversation_id: Optional[str] = None


class QueryExecutionRequest(BaseModel):
    """Request to execute a generated query"""
    sql: str
    connection_id: str
    max_rows: int = Field(default=1000, le=10000)
    timeout_seconds: int = Field(default=30, le=120)


class QueryExecutionResponse(BaseModel):
    """Response from query execution"""
    columns: list[str]
    rows: list[list[Any]]
    row_count: int
    execution_time_ms: float
    truncated: bool = False


# Chat Models

class ChatMessage(BaseModel):
    """A single message in a chat conversation"""
    id: str
    role: MessageRole
    content: str
    timestamp: datetime
    sql: Optional[str] = None
    query_result: Optional[dict[str, Any]] = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ChatSession(BaseModel):
    """A chat session with Kodee"""
    id: str
    user_id: Optional[str] = None
    connection_id: Optional[str] = None
    model_id: Optional[str] = None
    title: Optional[str] = None
    messages: list[ChatMessage] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    context: Optional[SchemaContext] = None


class ChatRequest(BaseModel):
    """Request for chat interaction"""
    message: str
    session_id: Optional[str] = None
    connection_id: Optional[str] = None
    model_id: Optional[str] = None
    execute_query: bool = True  # Whether to execute generated SQL
    include_visualization: bool = False


class ChatResponse(BaseModel):
    """Response from chat interaction"""
    session_id: str
    message_id: str
    response: str
    sql: Optional[str] = None
    query_result: Optional[QueryExecutionResponse] = None
    visualization_suggestion: Optional[dict[str, Any]] = None
    suggestions: list[str] = Field(default_factory=list)
    follow_up_questions: list[str] = Field(default_factory=list)


# Query History

class QueryHistoryItem(BaseModel):
    """A record of a previous query"""
    id: str
    user_id: Optional[str] = None
    question: str
    sql: str
    connection_id: str
    model_id: Optional[str] = None
    executed: bool = False
    execution_successful: Optional[bool] = None
    row_count: Optional[int] = None
    created_at: datetime


class QueryHistoryResponse(BaseModel):
    """Response with query history"""
    items: list[QueryHistoryItem]
    total: int
    page: int
    page_size: int


# Validation Request

class SQLValidationRequest(BaseModel):
    """Request to validate SQL"""
    sql: str
    connection_id: Optional[str] = None
    strict: bool = True  # Strict mode blocks more operations


# Feedback Models

class QueryFeedback(BaseModel):
    """User feedback on a generated query"""
    query_id: str
    rating: int = Field(..., ge=1, le=5)
    correct_sql: Optional[str] = None
    comments: Optional[str] = None


# Constants

BLOCKED_KEYWORDS = [
    "DROP", "DELETE", "TRUNCATE", "ALTER", "CREATE", "INSERT", "UPDATE",
    "GRANT", "REVOKE", "EXEC", "EXECUTE", "xp_", "sp_", "--", "/*", "*/"
]

AGGREGATION_KEYWORDS = [
    "total", "sum", "count", "average", "avg", "mean", "max", "maximum",
    "min", "minimum", "how many", "number of", "percentage", "percent"
]

COMPARISON_KEYWORDS = [
    "compare", "versus", "vs", "against", "difference", "between", "more than",
    "less than", "greater", "higher", "lower", "top", "bottom", "best", "worst"
]

TIME_KEYWORDS = [
    "today", "yesterday", "this week", "last week", "this month", "last month",
    "this year", "last year", "ytd", "mtd", "qtd", "quarter", "year over year",
    "yoy", "mom", "trend", "over time", "by date", "by month", "by year"
]


# Prompt Templates

SYSTEM_PROMPT_TEMPLATE = """You are Kodee, an AI assistant specialized in converting natural language questions into SQL queries.

You have access to the following database schema:

{schema_context}

Guidelines:
1. Generate only SELECT queries - no data modification
2. Use proper table aliases and quote identifiers appropriately for {dialect}
3. Include appropriate JOINs based on the relationships defined
4. Apply reasonable LIMIT clauses to prevent excessive data retrieval
5. Use the semantic model measures and dimensions when available
6. Format SQL for readability with proper indentation
7. Explain your query in plain language

Respond in JSON format with:
- sql: The generated SQL query
- explanation: Plain language explanation of what the query does
- confidence: Your confidence level (0-1) in the query being correct
- tables_used: List of tables referenced
- columns_used: List of columns referenced
"""

FEW_SHOT_EXAMPLES = [
    {
        "question": "What are the top 10 products by sales?",
        "sql": """SELECT p.product_name, SUM(s.amount) as total_sales
FROM products p
JOIN sales s ON p.product_id = s.product_id
GROUP BY p.product_name
ORDER BY total_sales DESC
LIMIT 10""",
        "explanation": "This query retrieves the top 10 products ranked by their total sales amount, joining the products and sales tables."
    },
    {
        "question": "How many orders were placed last month?",
        "sql": """SELECT COUNT(*) as order_count
FROM orders
WHERE order_date >= DATE_TRUNC('month', CURRENT_DATE - INTERVAL '1 month')
  AND order_date < DATE_TRUNC('month', CURRENT_DATE)""",
        "explanation": "This query counts all orders placed in the previous calendar month."
    },
    {
        "question": "Show me sales by region for this year",
        "sql": """SELECT r.region_name, SUM(s.amount) as total_sales
FROM sales s
JOIN regions r ON s.region_id = r.region_id
WHERE s.sale_date >= DATE_TRUNC('year', CURRENT_DATE)
GROUP BY r.region_name
ORDER BY total_sales DESC""",
        "explanation": "This query shows total sales grouped by region for the current year to date."
    }
]

File: frontend/src/types/kodee.ts
/**
 * Kodee NL-to-SQL Types
 *
 * TypeScript types for the Kodee natural language to SQL engine.
 */

// Enums

export type QueryIntent =
  | 'select'
  | 'aggregate'
  | 'compare'
  | 'trend'
  | 'top_n'
  | 'filter'
  | 'join'
  | 'unknown';

export type QueryComplexity = 'simple' | 'moderate' | 'complex';

export type MessageRole = 'user' | 'assistant' | 'system';

export type ValidationStatus = 'valid' | 'warning' | 'invalid' | 'blocked';

// Schema Context Types

export interface ColumnInfo {
  name: string;
  data_type: string;
  nullable: boolean;
  description?: string;
  sample_values: unknown[];
  is_primary_key: boolean;
  is_foreign_key: boolean;
  foreign_key_ref?: string;
}

export interface TableInfo {
  name: string;
  schema_name: string;
  description?: string;
  columns: ColumnInfo[];
  row_count?: number;
  aliases: string[];
}

export interface RelationshipInfo {
  from_table: string;
  from_column: string;
  to_table: string;
  to_column: string;
  relationship_type: string;
}

export interface MeasureInfo {
  id: string;
  name: string;
  description?: string;
  expression: string;
  table: string;
  column: string;
  aggregate: string;
}

export interface DimensionInfo {
  id: string;
  name: string;
  description?: string;
  table: string;
  column: string;
  hierarchy?: string[];
}

export interface SchemaContext {
  connection_id: string;
  model_id?: string;
  model_name?: string;
  tables: TableInfo[];
  relationships: RelationshipInfo[];
  measures: MeasureInfo[];
  dimensions: DimensionInfo[];
  dialect: string;
  additional_context?: string;
}

// Query Types

export interface NLQueryRequest {
  question: string;
  connection_id?: string;
  model_id?: string;
  conversation_id?: string;
  include_explanation?: boolean;
  max_rows?: number;
  timeout_seconds?: number;
}

export interface QueryValidation {
  status: ValidationStatus;
  messages: string[];
  blocked_keywords: string[];
  estimated_complexity?: QueryComplexity;
  estimated_rows?: number;
}

export interface NLQueryResponse {
  sql: string;
  explanation: string;
  confidence: number;
  intent: QueryIntent;
  complexity: QueryComplexity;
  validation: QueryValidation;
  tables_used: string[];
  columns_used: string[];
  alternatives: string[];
  follow_up_questions: string[];
  execution_time_ms?: number;
  conversation_id?: string;
}

export interface QueryExecutionRequest {
  sql: string;
  connection_id: string;
  max_rows?: number;
  timeout_seconds?: number;
}

export interface QueryExecutionResponse {
  columns: string[];
  rows: unknown[][];
  row_count: number;
  execution_time_ms: number;
  truncated: boolean;
}

// Chat Types

export interface ChatMessage {
  id: string;
  role: MessageRole;
  content: string;
  timestamp: string;
  sql?: string;
  query_result?: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export interface ChatSession {
  id: string;
  user_id?: string;
  connection_id?: string;
  model_id?: string;
  title?: string;
  messages: ChatMessage[];
  created_at: string;
  updated_at: string;
  context?: SchemaContext;
}

export interface ChatRequest {
  message: string;
  session_id?: string;
  connection_id?: string;
  model_id?: string;
  execute_query?: boolean;
  include_visualization?: boolean;
}

export interface ChatResponse {
  session_id: string;
  message_id: string;
  response: string;
  sql?: string;
  query_result?: QueryExecutionResponse;
  visualization_suggestion?: Record<string, unknown>;
  suggestions: string[];
  follow_up_questions: string[];
}

// History Types

export interface QueryHistoryItem {
  id: string;
  user_id?: string;
  question: string;
  sql: string;
  connection_id: string;
  model_id?: string;
  executed: boolean;
  execution_successful?: boolean;
  row_count?: number;
  created_at: string;
}

export interface QueryHistoryResponse {
  items: QueryHistoryItem[];
  total: number;
  page: number;
  page_size: number;
}

// Validation Types

export interface SQLValidationRequest {
  sql: string;
  connection_id?: string;
  strict?: boolean;
}

// Feedback Types

export interface QueryFeedback {
  query_id: string;
  rating: number;
  correct_sql?: string;
  comments?: string;
}

// Example Types

export interface QueryExample {
  question: string;
  sql: string;
  explanation: string;
}

export interface QueryCategory {
  name: string;
  questions: string[];
}

export interface QueryExamplesResponse {
  examples: QueryExample[];
  categories: QueryCategory[];
}

// Suggestions Types

export interface QuerySuggestionsResponse {
  suggestions: string[];
  schema_summary: {
    tables: number;
    measures: number;
    dimensions: number;
  };
}

// Constants

export const INTENT_LABELS: Record<QueryIntent, string> = {
  select: 'Data Selection',
  aggregate: 'Aggregation',
  compare: 'Comparison',
  trend: 'Trend Analysis',
  top_n: 'Top N',
  filter: 'Filtering',
  join: 'Join',
  unknown: 'Unknown',
};

export const COMPLEXITY_LABELS: Record<QueryComplexity, string> = {
  simple: 'Simple',
  moderate: 'Moderate',
  complex: 'Complex',
};

export const COMPLEXITY_COLORS: Record<QueryComplexity, string> = {
  simple: 'green',
  moderate: 'yellow',
  complex: 'red',
};

export const VALIDATION_STATUS_COLORS: Record<ValidationStatus, string> = {
  valid: 'green',
  warning: 'yellow',
  invalid: 'red',
  blocked: 'red',
};

// Sample Questions

export const SAMPLE_QUESTIONS = [
  'What are the top 10 products by sales?',
  'Show me total revenue by month',
  'How many orders were placed last quarter?',
  'Compare sales this year vs last year',
  'Which customers have the highest order value?',
  'Show me the trend of new customers over time',
  'What is the average order value by region?',
  'List products with low inventory',
];

// Helper Functions

export function formatConfidence(confidence: number): string {
  return `${Math.round(confidence * 100)}%`;
}

export function getConfidenceColor(confidence: number): string {
  if (confidence >= 0.8) return 'green';
  if (confidence >= 0.6) return 'yellow';
  return 'red';
}

export function createDefaultQueryRequest(
  question: string,
  connectionId?: string,
  modelId?: string
): NLQueryRequest {
  return {
    question,
    connection_id: connectionId,
    model_id: modelId,
    include_explanation: true,
    max_rows: 1000,
    timeout_seconds: 30,
  };
}

export function createDefaultChatRequest(
  message: string,
  sessionId?: string,
  connectionId?: string
): ChatRequest {
  return {
    message,
    session_id: sessionId,
    connection_id: connectionId,
    execute_query: true,
    include_visualization: false,
  };
}
